src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c

```diff
@@ -138,7 +157,15 @@ ARGS_NON_NULL_ALL static int32_t ReinitializeCipher(CipherCtx* ctx)
     {
         jbyteArray ivBytes = make_java_byte_array(env, ctx->ivLength);
         (*env)->SetByteArrayRegion(env, ivBytes, 0, ctx->ivLength, (jbyte*)ctx->iv);
-        if (HasTag(ctx->type))
+
+        // ChaChaPoly1305 does not allow configuration of the tag length and
+        // we don't permit user-defined counters, so we can use IvParameterSpec
+        // for ChaChaPoly1305.
+        if (strcmp("ChaCha20/Poly1305/NoPadding", ctx->type->name) == 0)
```

Comment:
This feels a little gross to me, happy to change this if someone has a better idea.

ChaChaPoly1305 does have an authentication tag, but that algorithm forbids a tag length other than 16 bytes. Because of that, ChaChaPoly1305 on Android can use the `IvParameterSpec`.

Reply:
I did a bit of refactoring here and I'm happier with where it ended up.